engine/consensus/dkg/reactor_engine.go

```diff
@@ -229,42 +229,59 @@ func (e *ReactorEngine) startDKGForEpoch(currentEpochCounter uint64, first *flow
+// of the current epoch, this guarantees that when we reach the end of the current epoch
+// we will either have a usable beacon key (successful DKG) or a DKG failure end state
+// stored, so we can safely fall back to using our staking key.
 //
-func (e *ReactorEngine) handleEpochCommittedPhaseStarted(currentEpochCounter uint64) {
+// CAUTION: This function is not safe for concurrent use. This is not enforced within
+// the ReactorEngine - instead we rely on the protocol event emission being single-threaded
+//
+func (e *ReactorEngine) handleEpochCommittedPhaseStarted(currentEpochCounter uint64, firstBlock *flow.Header) {
 
-	// the epoch counter the DKG just completed is for
-	dkgEpochCounter := currentEpochCounter + 1
+	// the DKG we have just completed produces keys that we will use in the next epoch
+	nextEpochCounter := currentEpochCounter + 1
 
 	log := e.log.With().
 		Uint64("cur_epoch", currentEpochCounter). // the epoch we are in the middle of
-		Uint64("dkg_for_epoch", dkgEpochCounter). // the epoch the just-finished DKG was preparing for
+		Uint64("next_epoch", nextEpochCounter).   // the epoch the just-finished DKG was preparing for
 		Logger()
 
-	// TODO - first check whether we've already stored the dkg end state, since
-	// we need to handle multiple calls to this method
+	// Check whether we have already set the end state for this DKG.
+	// This can happen if the DKG failed locally, if we failed to generate
+	// a local private beacon key, or if we crashed while performing this
+	// check previously.
+	endState, err := e.dkgState.GetDKGEndState(nextEpochCounter)
+	if err == nil {
+		log.Warn().Msgf("already set dkg end state when checking beacon key consistency: %s", endState.String())
+		return
+	}
 
-	// TODO execute in transaction scope of protocol event emitter, need to use
-	// lower-level storage access here (since State doesn't allow tx injection)
-	nextDKG, err := e.State.Final().Epochs().Next().DKG()
+	// Since epoch phase transitions are emitted when the first block of the new
+	// phase is finalized, the block's snapshot is guaranteed to already be
+	// accessible in the protocol state at this point (even though the Badger
+	// transaction finalizing the block has not been committed yet).
+	nextDKG, err := e.State.AtBlockID(firstBlock.ID()).Epochs().Next().DKG()
 	if err != nil {
 		// CAUTION: this should never happen, indicates a storage failure or corruption
 		log.Fatal().Err(err).Msg("checking beacon key consistency: could not retrieve next DKG info")
 		return
 	}
 
-	myBeaconPrivKey, err := e.dkgState.RetrieveMyBeaconPrivateKey(currentEpochCounter + 1)
+	myBeaconPrivKey, err := e.dkgState.RetrieveMyBeaconPrivateKey(nextEpochCounter)
 	if errors.Is(err, storage.ErrNotFound) {
 		log.Warn().Msg("checking beacon key consistency: no key found")
-		err := e.dkgState.SetDKGEndState(dkgEpochCounter, flow.DKGEndStateNoKey)
+		err := e.dkgState.SetDKGEndState(nextEpochCounter, flow.DKGEndStateNoKey)
```

Comment:
Aren't we overwriting the DKG end state here? For instance, in the case where there was a local failure, DKG end state is set to  `DKGEndStateDKGFailure` and no key is stored. 

We could : 
- avoid setting the DKG end state here. The state might stay in its default state `DKGEndStateUnknown` but that should be fine.
- OR read the state first and overwrite only if the state is `DKGEndStateUnknown`.
- OR merge `DKGEndStateDKGFailure` and `DKGEndStateNoKey` cases but then we would lose the detailed state type.

Reply:
> OR read the state first and overwrite only if the state is DKGEndStateUnknown.

This is what we do:
https://github.com/onflow/flow-go/blob/bdfe4b7ded8fd198c5cc821079c4ae4c499a5d8c/engine/consensus/dkg/reactor_engine.go#L256-L264

If we have a local DKG failure, we will store (`key=nil,end_state=dkg_failure`). If we restart during the DKG, we will store (`key=nil,end_state=no_key`).

This test case captures this scenario:
https://github.com/onflow/flow-go/blob/562dc57f75b674272289cd0b68d69df001142a11/engine/consensus/dkg/reactor_engine_test.go#L418

Reply:
You're right, I missed the check for the nil error